File: 200.number-of-islands.ts
/*
 * @lc app=leetcode id=200 lang=typescript
 *
 * [200] Number of Islands
 */

// @lc code=start
function numIslands(grid: string[][]): number {
  let count: number = 0;
  function dfs(grid: string[][], r: number, c: number): void {
    if (
      Math.min(r, c) < 0 ||
      r >= grid.length ||
      c >= grid[0].length ||
      grid[r][c] == "0"
    ) {
      return;
    }
    grid[r][c] = "0";
    dfs(grid, r + 1, c);
    dfs(grid, r - 1, c);
    dfs(grid, r, c + 1);
    dfs(grid, r, c - 1);
  }

  for (let r = 0; r < grid.length; r++) {
    for (let c = 0; c < grid[0].length; c++) {
      if (grid[r][c] == "1") {
        count++;
        dfs(grid, r, c);
      }
    }
  }

  return count;
}

// @lc code=end

File: 200.Number-of-Islands.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        res = 0

        def dfs(x, y):
            grid[x][y] = "-1"
            if x > 0 and grid[x-1][y] == "1":
                dfs(x-1, y)
            if x < len(grid) - 1 and grid[x+1][y] == "1":
                dfs(x+1, y)
            if y > 0 and grid[x][y-1] == "1":
                dfs(x, y-1)
            if y < len(grid[0]) - 1 and grid[x][y+1] == "1":
                dfs(x, y+1)

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    res += 1
                    dfs(i, j)

        return res
